engine/runtime/src/main/java/org/enso/interpreter/node/BinaryOperatorNode.java

```diff
@@ -0,0 +1,207 @@
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.NeverDefault;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.node.BinaryOperatorNodeFactory.DoThatConversionNodeGen;
+import org.enso.interpreter.node.callable.InteropConversionCallNode;
+import org.enso.interpreter.node.callable.dispatch.InvokeFunctionNode;
+import org.enso.interpreter.node.expression.builtin.meta.TypeOfNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedConversion;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.EnsoMultiValue;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.data.atom.Atom;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+import org.enso.interpreter.runtime.state.State;
+
+final class BinaryOperatorNode extends ExpressionNode {
+  private @Child ExpressionNode left;
+  private @Child ExpressionNode right;
+  private @Child ExpressionNode body;
+  private @Child DoThatConversionNode thatConversion;
+
+  BinaryOperatorNode(ExpressionNode left, ExpressionNode right, ExpressionNode body) {
+    this.body = body;
+    this.left = left;
+    this.right = right;
+  }
+
+  @Override
+  public Object executeGeneric(VirtualFrame frame) {
+    left.executeGeneric(frame);
+    try {
+      right.executeGeneric(frame);
+    } catch (PanicException ex) {
+      var ctx = EnsoContext.get(this);
+      if (ex.getPayload() instanceof Atom error && ctx.getBuiltins().error().isTypeError(error)) {
+        if (getRootNode() instanceof MethodRootNode methodRoot
+            && ex.getLocation() != null
+            && ex.getLocation().getRootNode() == methodRoot) {
+          if (thatConversion == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            thatConversion = insert(DoThatConversionNode.create());
+          }
+          var name = methodRoot.getMethodName();
+          var args = Function.ArgumentsHelper.getPositionalArguments(frame.getArguments());
+
+          var result = thatConversion.executeThatConversion(frame, name, args[0], args[1]);
+          if (result != null) {
+            return result;
+          } else {
+            throw ex;
+          }
+        }
+      }
+      throw ex;
+    }
+    return body.executeGeneric(frame);
+  }
+
+  /**
+   * A node that checks the type of {@code that} argument and performs conversion of {@code self} to
+   * {@code that} type, if it contains the requested symbol implementation.
```

Comment:
IMHO this description is a bit imprecise. It only speaks about it converting `self` to `that.type` but, it does not say anything about actually _executing_ the operation.

```suggestion
   * A node that checks the type of {@code that} argument, performs conversion of {@code self} to
   * {@code that} type, if it contains the requested symbol implementation, and executes that method instead of the original operator.
```